pkg/webhook/kafkacluster_validator.go

```diff
@@ -0,0 +1,73 @@
+// Copyright © 2019 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	banzaicloudv1beta1 "github.com/banzaicloud/koperator/api/v1beta1"
+)
+
+func (s *webhookServer) validateKafkaCluster(kafkaClusterNew *banzaicloudv1beta1.KafkaCluster) *admissionv1.AdmissionResponse {
+	ctx := context.Background()
+	// get the Old kafkaCluster CR
+	kafkaClusterSpecOld := banzaicloudv1beta1.KafkaCluster{}
+	err := s.client.Get(ctx, client.ObjectKey{Name: kafkaClusterNew.GetName(), Namespace: kafkaClusterNew.GetNamespace()}, &kafkaClusterSpecOld)
+	if err != nil {
+		// New kafkaCluster has been added thus no need to check storage removal
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		log.Error(err, "couldn't get KafkaCluster custom resource")
+		return notAllowed("API failure while retrieving KafkaCluster CR, please try again", metav1.StatusReasonInternalError)
+	}
+
+	return checkBrokerStorageRemoval(&kafkaClusterSpecOld.Spec, &kafkaClusterNew.Spec)
+}
+
+// checkBrokerStorageRemoval checks if there is any broker storage which has been removed. If yes, admission will be rejected
+func checkBrokerStorageRemoval(kafkaClusterSpecOld, kafkaClusterSpecNew *banzaicloudv1beta1.KafkaClusterSpec) *admissionv1.AdmissionResponse {
```

Comment:
The logic implemented in this function doesn't take into account brokers for which a downscale is in progress. The volumes used by that broker should not be allowed to be removed until the broker is gracefully downscaled. So brokers from the `status` field that are in either pending or running downscale needs to be taken into account as well beside the list of brokers in the spec.